common/ast/src/parser/query.rs

```diff
@@ -305,3 +271,68 @@ pub fn order_by_expr(i: Input) -> IResult<OrderByExpr> {
         },
     )(i)
 }
+
+pub fn set_expr(i: Input) -> IResult<SetExpr> {
+    let select = map(
+        rule! {
+             SELECT ~ DISTINCT? ~ #comma_separated_list1(select_target)
+                ~ ( FROM ~ ^#comma_separated_list1(table_reference) )?
+                ~ ( WHERE ~ ^#expr )?
+                ~ ( GROUP ~ ^BY ~ ^#comma_separated_list1(expr) )?
+                ~ ( HAVING ~ ^#expr )?
+        },
+        |(
+            _select,
+            opt_distinct,
+            select_list,
+            opt_from_block,
+            opt_where_block,
+            opt_group_by_block,
+            opt_having_block,
+        )| {
+            SetExpr::Select(Box::new(SelectStmt {
+                span: &[],
+                distinct: opt_distinct.is_some(),
+                select_list,
+                from: opt_from_block
+                    .map(|(_, table_refs)| table_refs)
+                    .unwrap_or_default(),
+                selection: opt_where_block.map(|(_, selection)| selection),
+                group_by: opt_group_by_block
+                    .map(|(_, _, group_by)| group_by)
+                    .unwrap_or_default(),
+                having: opt_having_block.map(|(_, having)| having),
+            }))
+        },
+    );
+
+    map(
+        rule!(#select ~ (UNION | EXCEPT | INTERSECT)? ~ ALL? ~ #set_expr?),
+        |(select, set_operator, all_kind, set_expr)| {
```

Comment:
There is execution precedence of set operators.

For example, 
```sql
(select * from t union select * from t1) intersect select * from t2
```

The SQL will be interpreted into:
```
   ∩
  / \
 ∪  t2
/ \
t t1
```

So the rule should be like `#set_expr ~ SET_OP ~ #set_expr`.

Reply:
Yes, I have thought the case and wanted to test the current rule if satisfy it, but I got the following error
```
mysql> (select * from t1 intersect select * from t2) intersect select * from t1;
ERROR 1105 (HY000): Code: 1005, displayText = error:
  --> SQL:1:1
  |
1 | (select * from t1 intersect select * from t2) intersect select * from t1
  | ^ expected `EXPLAIN`, `SELECT`, `SHOW`, `CREATE`, `DROP`, `ALTER`, or 10 more ...

.
```



> #set_expr ~ SET_OP ~ #set_expr

This should be the rule of `SetOperation`, for `SetExpr`, the rule may be `#select | #set_expr`, but this will cause dead loop.

Reply:
```ebnf
paren_select := ( '(', paren_select, ')' ) | select;

set_operation := set_expr, SET_OP, set_expr;

set_expr :=  set_operation | subquery | paren_select;
```

You may need this.